### ISSUE TITLE
FirstAsync() works, Take(1).ToListAsync() times out, Take(1).FirstAsync() throws

### ISSUE THREAD

**Issue by joelverhagen** [dotnet/efcore#14324]
I have three variants of a query. Each query should do the same thing but one variant works, one variant times out, and one variant throws an exception about an invalid column name.

From a high level, the common part of the EF query is:
```csharp
var query = entities
    .Packages
    .Include(x => x.PackageRegistration)
    .Include(x => x.CatalogPackage)
    .Where(x => x.CatalogPackage != null)
    .Where(x => x.CatalogPackage.LastCommitTimestamp > start && x.CatalogPackage.LastCommitTimestamp <= end)
    .OrderBy(x => x.CatalogPackage.LastCommitTimestamp);
```

The three variants are:
```csharp
await query.FirstAsync();           // works
await query.Take(1).ToListAsync()); // times out
await query.Take(1).FirstAsync();   // fails with invalid query
```

For the first two variants, the query produced is exactly the same except one case has `TOP(1)` and the other has `TOP(@__p_2)` where `@__p_2` is 1.

The timing out query works just fine when executing against SQL Server directly via SQL Server Management Studio. I originally found this because I had a `Take(5000)` which was timing out in EF but not timing out when I executed that very same query directly.

The exceptions are:

```
System.Data.SqlClient.SqlException (0x80131904): Timeout expired.  The timeout period elapsed prior to completion of the operation or the server is not responding. ---> System.ComponentModel.Win32Exception (258): The wait operation timed out
   at System.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__122_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteAsync(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.BufferlessMoveNext(DbContext _, Boolean buffer, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.MoveNextCore(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\Select.cs:line 106
   at System.Linq.AsyncEnumerable.AsyncIterator`1.MoveNext(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\AsyncIterator.cs:line 98
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.Aggregate_[TSource,TAccumulate,TResult](IAsyncEnumerable`1 source, TAccumulate seed, Func`3 accumulator, Func`2 resultSelector, CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\Aggregate.cs:line 120
   at ConsoleApp1.Program.ExecuteAsync(String connectionString, Func`2 execute) in D:\trash\2019-01-04\ConsoleApp1\ConsoleApp1\Program.cs:line 50
ClientConnectionId:7848828d-0005-457a-aeec-e30027d2f179
Error Number:-2,State:0,Class:11
```

```
System.Data.SqlClient.SqlException (0x80131904): Invalid column name 'c'.
   at System.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__122_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteAsync(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.BufferlessMoveNext(DbContext _, Boolean buffer, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.MoveNextCore(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\Select.cs:line 106
   at System.Linq.AsyncEnumerable.AsyncIterator`1.MoveNext(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\AsyncIterator.cs:line 98
   at System.Linq.AsyncEnumerable.First_[TSource](IAsyncEnumerable`1 source, CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\First.cs:line 125
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.TaskResultAsyncEnumerable`1.Enumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.MoveNextCore(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\Select.cs:line 106
   at System.Linq.AsyncEnumerable.AsyncIterator`1.MoveNext(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\AsyncIterator.cs:line 98
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteSingletonAsyncQuery[TResult](QueryContext queryContext, Func`2 compiledQuery, IDiagnosticsLogger`1 logger, Type contextType)
   at ConsoleApp1.Program.ExecuteAsync(String connectionString, Func`2 execute) in D:\trash\2019-01-04\ConsoleApp1\ConsoleApp1\Program.cs:line 51
ClientConnectionId:b67d62c3-57a3-4c25-a128-e3b51a00ad7e
Error Number:207,State:1,Class:16
```

### Steps to reproduce

I tried to come up with a minimal repro for the hang, but I could not on LocalDB. My initial repro was on a populated Azure SQL database. Feel free to reach out to me (jver, Microsoft employee) if you want that repro.

I was able to make a minimal repro on LocalDB for the invalid query... so perhaps there are two underlying issues (one for the bad query, one for the timeout):
https://gist.github.com/joelverhagen/88560062766acbe8adc290c37fea0b23

Try a connection string like this, assuming you have LocalDB on your machine.

```
Data Source=(localdb)\mssqllocaldb; Initial Catalog=ReproForJoel; Integrated Security=True; MultipleActiveResultSets=True
```

### Further technical details
EF Core version: 2.2.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017 15.9.5 Preview 1.0
Target framework: netcoreapp2.2
.NET Core SDK: 2.2.101

**Comment by joelverhagen** [comment]
I found that removing any part of the EF query or adding `.Where(x.PackageRegistration != null)` (which is pretty much meaningless since a foreign constraint means this will always be true) makes the "timeout" case work.

<pre>
var query = entities
    .Packages
    .Include(x => x.PackageRegistration)
    .Include(x => x.CatalogPackage)
    <b>.Where(x => x.PackageRegistration != null)</b>
    .Where(x => x.CatalogPackage != null)
    .Where(x => x.CatalogPackage.LastCommitTimestamp > start && x.CatalogPackage.LastCommitTimestamp <= end)
    .OrderBy(x => x.CatalogPackage.LastCommitTimestamp);
</pre>

There seems to be something special about the original query that makes it hang.

**Comment by smitpatel** [comment]
Tried the gist on 3.1 release and output is
```
Starting...
Success in 00:00:00.1009532.
Starting...
Success in 00:00:00.0505660.
Starting...
Success in 00:00:00.0100707.
```
Hence, invalid SQL issue is fixed in 3.1 for sure. Timeout issue may not occur in the absence of data.

@joelverhagen - Can you verify that it works as expected in all cases with EF Core 3.1

**Comment by smitpatel** [comment]
Closing as fixed. I have verified that it is working in 3.1. Since timeout issue is data dependent, we cannot test it accurately. If the issue persists past 3.1 feel free to reopen.
